File: static/js/core/Track.js
/* * * * * * * * * * * * * * * * * * * * * * * * * *
 *                                                 *
 *  Track class                                    *
 *                                                 *
 *  Track object from db with all its metadata     *
 *                                                 *
 *  track : {object} Raw track from database JSON  *
 *                                                 *
 * * * * * * * * * * * * * * * * * * * * * * * * * */

import MzkObject from './MzkObject.js'

class Track extends MzkObject{

    constructor(track) {
        super();
        this.updateMetadata(track);
    }

//  --------------------------------  PUBLIC METHODS  ---------------------------------  //

    /**
     * method : updateMetadata (public)
     * class  : Track
     * desc   : Update metadata contained in Track (UI) object
     * arg    : {object} track - Raw JSON track
     **/
    updateMetadata(track) {
        this.id = {
            track:          track.ID                ? track.ID                : "",
            album:          track.ALBUM.ID          ? track.ALBUM.ID          : "",
            artists:        this._getArtistsIDFromArtistsArray(track.ARTISTS)
        };
        this.title        = track.TITLE             ? track.TITLE             : "";
        this.year         = track.YEAR              ? track.YEAR              : "";
        this.composer     = track.COMPOSER          ? track.COMPOSER          : "";
        this.performer    = track.PERFORMER         ? track.PERFORMER         : "";
        this.track        = track.TRACK_NUMBER      ? track.TRACK_NUMBER      : "";
        this.trackTotal   = track.ALBUM.TOTAL_TRACK ? track.ALBUM.TOTAL_TRACK : "";
        this.disc         = track.DISC_NUMBER       ? track.DISC_NUMBER        : "";
        this.discTotal    = track.ALBUM.TOTAL_DISC  ? track.ALBUM.TOTAL_DISC  : "";
        this.bpm          = track.BPM               ? track.BPM               : "";
        this.lyrics       = track.LYRICS            ? track.LYRICS            : "";
        this.comment      = track.COMMENT           ? track.COMMENT           : "";
        this.bitRate      = track.BITRATE           ? track.BITRATE           : "";
        this.sampleRate   = track.SAMPLERATE        ? track.SAMPLERATE        : "";
        this.duration     = track.DURATION          ? track.DURATION          : "";
        this.size         = track.SIZE              ? track.SIZE              : "";
        this.lastModified = track.LAST_MODIFIED     ? track.LAST_MODIFIED     : "";
        this.album        = track.ALBUM.TITLE       ? track.ALBUM.TITLE       : "";
        this.genre        = track.GENRE             ? track.GENRE             : "";
        this.fileType     = track.FILE_TYPE         ? track.FILE_TYPE         : "";
        this.cover        = track.COVER             ? "../static/img/covers/" + track.COVER : "../static/img/utils/defaultcover.svg";
        this.artist       = this._getArtistFromArtistsArray(track.ARTISTS);
        this.albumArtist  = this._getArtistFromArtistsArray(track.ALBUM.ARTISTS);
        this.playCount    = track.PLAY_COUNTER      ? track.PLAY_COUNTER      : 0;
        this.fileName     = track.FILENAME          ? track.FILENAME          : "";
    }

//  --------------------------------  PRIVATE METHODS  --------------------------------  //

    /**
     * method : _getArtistsIDFromArtistsArray (private)
     * class  : Track
     * desc   : Extract artists IDs from JSON
     * arg    : {[object]} artists - Raw JSON array of objects
     * return : {[int]} - The artists ID array
     **/
    _getArtistsIDFromArtistsArray(artists) {
        if (artists === null || artists === undefined) { return ""; }

        let artistsID = [];
        for (let i = 0; i < artists.length; ++i) {
            artistsID.push(artists[i].ID);
        }

        return artistsID;
    }


    /**
     * method : _getArtistFromArtistsArray (private)
     * class  : Track
     * desc   : Create artists string from artist's names in [object]
     * arg    : {[object]} artists - Raw JSON array of objects
     * return : {string} The Artists concated string
     **/
    _getArtistFromArtistsArray(artists) {
        if (artists === null || artists === undefined) { return ""; }

        let artistsName = []; // Artists name array
        for (let i = 0; i < artists.length; ++i) {
            artistsName.push(artists[i].NAME);
        }

        artistsName.sort(); // In order to get artists alphabetically ordered

        let artist = ""; // Output string
        for (let i = 0; i < artistsName.length; ++i) {
            artist += artistsName[i];
            if (i < (artistsName.length - 1)) { artist += ", "; }
        }

        return artist;
    }

}

export default Track

File: app/track/track.py
import json
import zipfile

import os

import zlib
from django.contrib.auth.decorators import login_required
from django.http import JsonResponse
from django.utils.html import strip_tags

from app.history import addToHistory
from app.models import Track
from app.stats.stats import addToStats
from app.utils import errorCheckMessage, checkPermission


# Scan all the attributes of an MP3 track, and add it to base.
from app.wallet import checkListeningGain


def exportTrackInfo(track):
    if track.genre is not None:
        genre = track.genre.name
    else:
        genre = None

    artists = []
    for artist in track.artist.all():
        artists.append({
            'ID': artist.id,
            'NAME': artist.name,
        })

    album = {}
    if track.album is not None:
        albumArtists = []
        for artist in track.album.artist.all():
            albumArtists.append({
                'ID': artist.id,
                'NAME': artist.name,
            })
        album = {
            'ID': track.album.id,
            'TITLE': track.album.title,
            'TOTAL_DISC': track.album.totalDisc,
            'TOTAL_TRACK': track.album.totalTrack,
            'ARTISTS': albumArtists,
        }

    data = {
        'ID': track.id,
        'TITLE': track.title,
        'YEAR': track.year,
        'COMPOSER': track.composer,
        'PERFORMER': track.performer,
        'TRACK_NUMBER': track.number,
        'BPM': track.bpm,
        'LYRICS': track.lyrics,
        'COMMENT': track.comment,
        'BITRATE': track.bitRate,
        'SAMPLERATE': track.sampleRate,
        'DURATION': track.duration,
        'GENRE': genre,
        'FILE_TYPE': track.fileType.name,
        'DISC_NUMBER': track.discNumber,
        'SIZE': track.size,
        'LAST_MODIFIED': track.lastModified,
        'COVER': track.coverLocation,
        'ARTISTS': artists,
        'ALBUM': album,
        'PLAY_COUNTER': track.playCounter,
        'FILENAME': os.path.basename(track.location)
    }
    return data


# Get all the information about a track
@login_required(redirect_field_name='login.html', login_url='app:login')
def getTracksDetailedInfo(request):
    if request.method == 'POST':
        response = json.loads(request.body)
        user = request.user
        if checkPermission(["PLAY"], user):
            if 'TRACK_ID' in response:
                trackIds = response['TRACK_ID']
                trackInfo = []
                for trackId in trackIds:
                    if Track.objects.filter(id=trackId).count() == 1:
                        trackInfo.append(exportTrackInfo(Track.objects.get(id=trackId)))
                    else:
                        data = errorCheckMessage(False, "dbError")
                        return JsonResponse(data)
                data = {**dict({'RESULT': trackInfo}), ** errorCheckMessage(True, None)}
            else:
                data = errorCheckMessage(False, "badFormat")
        else:
            data = errorCheckMessage(False, "permissionError")
    else:
        data = errorCheckMessage(False, "badRequest")
    return JsonResponse(data)


# Return the link to a track with a track id
@login_required(redirect_field_name='login.html', login_url='app:login')
def getTrackPath(request):
    if request.method == 'POST':
        response = json.loads(request.body)
        user = request.user
        if checkPermission(["PLAY"], user):
            # Checking JSON keys
            if 'TRACK_ID' in response and 'PREVIOUS' in response and 'LAST_TRACK_PATH' in response \
                    and 'TRACK_PERCENTAGE' in response:
                trackId = strip_tags(response['TRACK_ID'])
                # Getting the track asked
                if Track.objects.filter(id=trackId).count() == 1:
                    track = Track.objects.get(id=trackId)
                    # If we don't ask a previous track
                    if not bool(response['PREVIOUS']):
                        # Adding the current track to the history
                        # Removing the first 2 chars
                        previousTrackPath = strip_tags(response['LAST_TRACK_PATH'])[2:]
                        # If the previous track exists
                        if Track.objects.filter(location=previousTrackPath).count() == 1:
                            listeningPercentage = float(strip_tags(response['TRACK_PERCENTAGE']))
                            previousTrack = Track.objects.get(location=previousTrackPath)
                            checkListeningGain(previousTrack, user)
                            # Adding to stats if the user has listened more than 15% of the song
                            if listeningPercentage > 5:
                                previousTrack.playCounter += 1
                                previousTrack.save()
                                addToStats(previousTrack, listeningPercentage, user)
                        addToHistory(track, user)

                    # Returning the asked song
                    data = {
                        'TRACK_PATH': track.location,
                    }
                    data = {**data, **errorCheckMessage(True, None)}
                else:
                    data = errorCheckMessage(False, "dbError")
            else:
                data = errorCheckMessage(False, "badFormat")
        else:
            data = errorCheckMessage(False, "permissionError")
    else:
        data = errorCheckMessage(False, "badRequest")
    return JsonResponse(data)


# Return the mood file with a given track id
@login_required(redirect_field_name='login.html', login_url='app:login')
def getMoodbar(request):
    if request.method == 'POST':
        response = json.loads(request.body)
        user = request.user
        if checkPermission(["PLAY"], user):
            if 'TRACK_ID' in response:
                trackID = response['TRACK_ID']
                if Track.objects.filter(id=trackID).count() == 1:
                    track = Track.objects.get(id=trackID)
                    data = {
                        'TRACK_MOOD': track.moodbar,
                    }
                    data = {**data, **errorCheckMessage(True, None)}
                else:
                    data = errorCheckMessage(False, "dbError")
            else:
                data = errorCheckMessage(False, "badFormat")
        else:
            data = errorCheckMessage(False, "permissionError")
    else:
        data = errorCheckMessage(False, "badRequest")
    return JsonResponse(data)


# Download the given song
@login_required(redirect_field_name='login.html', login_url='app:login')
def getDownloadLocation(request):
    if request.method == 'POST':
        response = json.loads(request.body)
        user = request.user
        if checkPermission(["DOWN"], user):
            if 'TRACK_ID' in response:
                trackId = strip_tags(response['TRACK_ID'])
                if Track.objects.filter(id=trackId).count() == 1:
                    track = Track.objects.get(id=trackId)
                    track.downloadCounter += 1
                    track.save()
                    data = {
                        'DOWNLOAD_PATH': track.location,
                    }
                    data = {**data, **errorCheckMessage(True, None)}
                else:
                    data = errorCheckMessage(False, "dbError")
            else:
                data = errorCheckMessage(False, "badFormat")
        else:
            data = errorCheckMessage(False, "permissionError")
    else:
        data = errorCheckMessage(False, "badRequest")
    return JsonResponse(data)


# Download a zip of different song
@login_required(redirect_field_name='login.html', login_url='app:login')
def multiTrackDownload(request):
    if request.method == "POST":
        response = json.loads(request.body)
        user = request.user
        if checkPermission(["DOWN"], user):
            if 'TRACKS_ID' in response:
                trackIds = response['TRACKS_ID']
                # TODO : create admin option max number of sound to download
                if len(trackIds) > 50:
                    return JsonResponse(errorCheckMessage(False, ""))
                locations = []

                # Getting tracks requested by the user
                for trackId in trackIds:
                    if Track.objects.filter(id=trackId).count() == 1:
                        track = Track.objects.get(id=trackId)
                        locations.append(track.location)
                tmp = ""
                for loc in locations:
                    tmp += loc
                archiveName = "ManaZeak-" + str(zlib.crc32(tmp.encode("ascii", "ignore"))) + ".zip"

                # Checking if the output folder for the zip exists
                if not os.path.isdir("/static/zip"):
                    try:
                        os.makedirs("/static/zip")
                    except OSError:
                        return JsonResponse(errorCheckMessage(False, "dirCreationError"))

                # Creating archive
                archiveName = os.path.join("/static/zip", archiveName)
                archive = zipfile.ZipFile(archiveName, 'w', zipfile.ZIP_DEFLATED)
                for location in locations:
                    archive.write(location, os.path.basename(location), compress_type=zipfile.ZIP_DEFLATED)

                data = {**{'DOWNLOAD_PATH': archiveName, }, **errorCheckMessage(True, None)}
            else:
                data = errorCheckMessage(False, "badFormat")
        else:
            data = errorCheckMessage(False, "permissionError")
    else:
        data = errorCheckMessage(False, "badRequest")
    return JsonResponse(data)
